### ISSUE TITLE
Cost calculation in Variational Autoencoder model is wrong

### ISSUE THREAD

**Issue by haoyangz** [tensorflow/models#2243]
### System information
- **What is the top-level directory of the model you are using**: autoencoder
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 14.04
- **TensorFlow installed from (source or binary)**: not relevant
- **TensorFlow version (use command below)**: not relevant
- **Bazel version (if compiling from source)**:not relevant
- **CUDA/cuDNN version**: not relevant
- **GPU model and memory**: not relevant
- **Exact command to reproduce**: not relevant

### Describe the problem
The [reconstruction error](https://github.com/tensorflow/models/blob/master/autoencoder/autoencoder_models/VariationalAutoencoder.py#L24) in variational autoencoder is calculated without specifying the "axis" argument, resulting in calculating the sum of reconstruction error across the whole batch. However, the [latent error](https://github.com/tensorflow/models/blob/master/autoencoder/autoencoder_models/VariationalAutoencoder.py#L25) is calculated with axis=1, resulting in a 1D vector representing the latent error for each sample in the batch. When we add them up and take the average to get the [cost](https://github.com/tensorflow/models/blob/master/autoencoder/autoencoder_models/VariationalAutoencoder.py#L26), the 0-d reconstruction error will get broadcast to each element of the latent error, resulting in a larger cost than it should be. The right way should be either mean + mean, or mean(vector + vector).

### Source code / logs
N/A

**Comment by masju** [comment]
Please unsubscribe

**Comment by haoyangz** [comment]
@jo8hua @snurkabill @nealwu @daviddao

**Comment by jo8hua** [comment]
@haoyangz 
I agree. I think if you removed the `axis=1` you could get rid of the call to `reduce_mean` all together. Feel free to make a PR.

**Comment by haoyangz** [comment]
@jo8hua That will results in the sum of cost, not the mean of cost (across a batch). So I think the best way is to add `axis=1` to reconstruction error calculation?

**Comment by jo8hua** [comment]
This is inconsequential (unless you are doing something like irregular batches). Right now I would say that the only issue is that the example is inconsistent.

**Comment by haoyangz** [comment]
Well I think the biggest problem of training against the sum of cost is that the training objective is now batch-size dependent. This means a larger batch size would require a smaller learning rate to get to the same weight upgrade, which will cause a lot of pain in hyper-parameter tuning. I will file a PR soon. Thanks!

**Comment by snurkabill** [comment]
@haoyangz I agree. Dependency between batch size and learning rate is obvious. This was practically simple implementation of scientific paper. It will be great if you provide some PR! Thank you :)

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.
